src/hotspot/os/linux/os_linux.cpp

```diff
@@ -585,22 +585,37 @@ void os::init_system_properties_values() {
     char *pslash;
     os::jvm_path(buf, bufsize);
 
-    // Found the full path to libjvm.so.
-    // Now cut the path to <java_home>/jre if we can.
-    pslash = strrchr(buf, '/');
-    if (pslash != nullptr) {
-      *pslash = '\0';            // Get rid of /libjvm.so.
-    }
+    // Found the full path to the binary. Note the code
+    // is based on the known JDK directory structure:
+    //
+    //   <jdk_path>/lib/{client|server|hotspot}/libjvm.so
+    //
+    // or
+    //
+    //   <jdk_path>/bin/<static_binary>
+#define JVM_LIB_NAME "libjvm.so"
     pslash = strrchr(buf, '/');
     if (pslash != nullptr) {
-      *pslash = '\0';            // Get rid of /{client|server|hotspot}.
+      if (strncmp(pslash + 1, JVM_LIB_NAME, strlen(JVM_LIB_NAME)) == 0) {
+        // Binary name is libjvm.so. Get rid of /libjvm.so.
+        *pslash = '\0';
+      }
+
+      // Get rid of /{client|server|hotspot}, if binary is libjvm.so.
+      // Or, cut off /<binary_name>.
```

Comment:
@jianglizhou This code is based on changes in the Hermetic Java repo, but I do not fully understand neither the comment nor what the purpose is. If you could explain this a bit I'd be grateful.

Reply:
The specific related commit in the hermetic Java branch is https://github.com/openjdk/leyden/commit/53aa8f0cf418ab5f435a4b9996c7754fb8505d4b. 

The change in os_linux.cpp here is to make sure  that the libjvm.so related path manipulation is conditionally done only. The check at line 599 looks for "/libjvm.so" substring, so we only chop off (`*pslash = `\0` at line 601) that part when necessary. In the static JDK case, there is no `libjvm.so` and the path string is `<jdk_path>/bin/javastatic`, which should not be affected. Otherwise, it could fail.

I found the code was not very easy to follow when running into problems and fixing for static support. So I added a bit more comments in the code here. The comment above about `/{client|server|hotspot}` was there originally. I think we no longer have those directories. We can cleanup that later, since it needs some more testing.

@magicus, thanks a lot for extracting/reworking/cleaning-up the static Java changes from the hermetic Java branch. That's a substantial amount of work!

I have one quick comment about the removal of `STATIC_LIB_EXCLUDE_OBJS` changes. Will post it as separate comment for the related code. 

I'll also look closely of the vm & jdk changes and compare those with the changes in the hermetic Java branch this week.